File: backend/src/schemas/notifications.py
"""
Schemas for notification channels, templates, recipients, rules, and delivery logs.
"""
from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class NotificationChannelBase(BaseModel):
    name: str = Field(..., max_length=255, description="Уникальное имя канала")
    type: str = Field(..., max_length=50, description="Тип канала (email, telegram, slack, sms и т.д.)")
    config: dict = Field(default_factory=dict, description="Конфигурация транспорта")
    enabled: bool = Field(default=True, description="Включен ли канал")
    status: str = Field(default="ok", max_length=32, description="Текущее состояние канала")
    concurrency_limit: Optional[int] = Field(None, ge=1, description="Ограничение параллельных отправок")
    retry_attempts: int = Field(default=3, ge=0, description="Количество ретраев")
    retry_interval_sec: int = Field(default=30, ge=0, description="Интервал между ретраями")
    timeout_sec: int = Field(default=10, ge=1, description="Таймаут отправки")
    is_primary: bool = Field(default=False, description="Флаг основного канала")


class NotificationChannelCreate(NotificationChannelBase):
    pass


class NotificationChannelUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=255)
    config: Optional[dict] = None
    enabled: Optional[bool] = None
    status: Optional[str] = Field(None, max_length=32)
    concurrency_limit: Optional[int] = Field(None, ge=1)
    retry_attempts: Optional[int] = Field(None, ge=0)
    retry_interval_sec: Optional[int] = Field(None, ge=0)
    timeout_sec: Optional[int] = Field(None, ge=1)
    is_primary: Optional[bool] = None


class NotificationChannelResponse(NotificationChannelBase):
    id: UUID
    test_at: Optional[datetime]
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        orm_mode = True


class NotificationTemplateBase(BaseModel):
    name: str = Field(..., max_length=255)
    locale: str = Field(default="en", max_length=5)
    subject: Optional[str] = Field(None, max_length=255)
    body: str = Field(..., description="Текст шаблона")
    variables: Optional[dict] = Field(None, description="Описание переменных")
    channel_id: Optional[UUID] = Field(None, description="Связанный канал")


class NotificationTemplateCreate(NotificationTemplateBase):
    pass


class NotificationTemplateUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=255)
    locale: Optional[str] = Field(None, max_length=5)
    subject: Optional[str] = Field(None, max_length=255)
    body: Optional[str] = None
    variables: Optional[dict] = None
    channel_id: Optional[UUID] = None


class NotificationTemplateResponse(NotificationTemplateBase):
    id: UUID
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        orm_mode = True


class NotificationRecipientBase(BaseModel):
    type: str = Field(..., max_length=50, description="Тип адресата (email, telegram, slack, sms)")
    address: str = Field(..., max_length=255, description="Адрес или идентификатор")
    silence_windows: Optional[dict] = Field(None, description="Окна тишины")
    status: str = Field(default="active", max_length=32, description="active | blocked | opt-out")


class NotificationRecipientCreate(NotificationRecipientBase):
    pass


class NotificationRecipientUpdate(BaseModel):
    address: Optional[str] = Field(None, max_length=255)
    silence_windows: Optional[dict] = None
    status: Optional[str] = Field(None, max_length=32)


class NotificationRecipientResponse(NotificationRecipientBase):
    id: UUID
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        orm_mode = True


class NotificationRuleBase(BaseModel):
    name: str = Field(..., max_length=255)
    enabled: bool = Field(default=True)
    severity_filter: Optional[dict] = Field(None, description="Фильтры по уровню")
    tag_filter: Optional[dict] = Field(None, description="Фильтры по тегам")
    host_filter: Optional[dict] = Field(None, description="Фильтры по хостам")
    failover_timeout_sec: int = Field(default=30, ge=0)
    silence_windows: Optional[dict] = Field(None, description="Окна тишины")
    rate_limit: Optional[dict] = Field(None, description="Лимиты отправки")
    dedup_window_sec: int = Field(default=0, ge=0)
    template_id: Optional[UUID] = None
    recipient_ids: List[UUID] = Field(default_factory=list)
    channel_ids: List[UUID] = Field(default_factory=list)
    test_channel_ids: Optional[List[UUID]] = None


class NotificationRuleCreate(NotificationRuleBase):
    pass


class NotificationRuleUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=255)
    enabled: Optional[bool] = None
    severity_filter: Optional[dict] = None
    tag_filter: Optional[dict] = None
    host_filter: Optional[dict] = None
    failover_timeout_sec: Optional[int] = Field(None, ge=0)
    silence_windows: Optional[dict] = None
    rate_limit: Optional[dict] = None
    dedup_window_sec: Optional[int] = Field(None, ge=0)
    template_id: Optional[UUID] = None
    recipient_ids: Optional[List[UUID]] = None
    channel_ids: Optional[List[UUID]] = None
    test_channel_ids: Optional[List[UUID]] = None


class NotificationRuleResponse(NotificationRuleBase):
    id: UUID
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        orm_mode = True


class DeliveryLogResponse(BaseModel):
    id: UUID
    event_id: str
    rule_id: Optional[UUID]
    channel_id: Optional[UUID]
    recipient_id: Optional[UUID]
    status: str
    attempt: int
    latency_ms: Optional[int]
    response_code: Optional[int]
    response_body: Optional[str]
    error_message: Optional[str]
    created_at: datetime

    class Config:
        orm_mode = True

File: mobile/src/api/notifications.ts
/**
 * Notifications API Module
 *
 * Provides API functions for managing notification rules, channels, templates, recipients, and delivery logs.
 * Follows patterns from frontend/src/api/notifications.ts adapted for React Native.
 *
 * Features:
 * - Full CRUD for notification rules, channels, templates, recipients
 * - Delivery log filtering and querying
 * - Test functionality for rules and channels
 */

import { apiClient } from './client';

// ============================================
// Type Definitions
// ============================================

export interface NotificationChannel {
  id: string;
  name: string;
  type: string;
  config: Record<string, unknown>;
  enabled: boolean;
  status: string;
  concurrency_limit?: number | null;
  retry_attempts: number;
  retry_interval_sec: number;
  timeout_sec: number;
  is_primary: boolean;
  test_at?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export type NotificationChannelCreate = Omit<NotificationChannel, 'id' | 'created_at' | 'updated_at' | 'test_at'>;
export type NotificationChannelUpdate = Partial<NotificationChannelCreate>;

export interface NotificationTemplate {
  id: string;
  name: string;
  locale: string;
  subject?: string | null;
  body: string;
  variables?: Record<string, unknown> | null;
  channel_id?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export type NotificationTemplateCreate = Omit<NotificationTemplate, 'id' | 'created_at' | 'updated_at'>;
export type NotificationTemplateUpdate = Partial<NotificationTemplateCreate>;

export interface NotificationRecipient {
  id: string;
  type: string;
  address: string;
  status: 'active' | 'blocked' | 'opt-out' | string;
  silence_windows?: Record<string, unknown> | null;
  created_at: string;
  updated_at?: string | null;
}

export type NotificationRecipientCreate = Omit<NotificationRecipient, 'id' | 'created_at' | 'updated_at'>;
export type NotificationRecipientUpdate = Partial<NotificationRecipientCreate>;

export interface NotificationRule {
  id: string;
  name: string;
  enabled: boolean;
  severity_filter?: Record<string, unknown> | null;
  tag_filter?: Record<string, unknown> | null;
  host_filter?: Record<string, unknown> | null;
  failover_timeout_sec: number;
  silence_windows?: Record<string, unknown> | null;
  rate_limit?: Record<string, unknown> | null;
  dedup_window_sec: number;
  template_id?: string | null;
  recipient_ids: string[];
  channel_ids: string[];
  test_channel_ids?: string[] | null;
  created_at: string;
  updated_at?: string | null;
}

export type NotificationRuleCreate = Omit<NotificationRule, 'id' | 'created_at' | 'updated_at'>;
export type NotificationRuleUpdate = Partial<NotificationRuleCreate>;

export interface RuleTestRequest {
  event_id?: string;
  severity?: string;
  tags?: Record<string, unknown>;
  host?: string;
  context?: Record<string, unknown>;
  subject?: string;
  body?: string;
}

export interface RuleTestResponse {
  status: string;
  event_id: string;
  tasks_enqueued: number;
}

export type DeliveryLogStatus =
  | 'success'
  | 'fail'
  | 'failover'
  | 'suppressed'
  | 'rate-limited'
  | 'deduped'
  | string;

export interface DeliveryLog {
  id: string;
  event_id: string;
  rule_id?: string | null;
  channel_id?: string | null;
  recipient_id?: string | null;
  status: DeliveryLogStatus;
  attempt: number;
  latency_ms?: number | null;
  response_code?: number | null;
  response_body?: string | null;
  error_message?: string | null;
  created_at: string;
}

export interface DeliveryLogFilters {
  rule_id?: string;
  channel_id?: string;
  recipient_id?: string;
  event_id?: string;
  statuses?: string[];
  created_from?: string;
  created_to?: string;
  limit?: number;
}

export interface ChannelTestRequest {
  recipient: string;
  subject?: string;
  body?: string;
  context?: Record<string, unknown>;
  use_celery?: boolean;
}

export interface ChannelTestResponse {
  status: string;
  event_id: string;
}

// ============================================
// API Functions
// ============================================

export const notificationsApi = {
  // Channels
  listChannels: async (): Promise<NotificationChannel[]> => {
    const response = await apiClient.get<NotificationChannel[]>('/api/notifications/channels');
    return response.data;
  },

  createChannel: async (data: NotificationChannelCreate): Promise<NotificationChannel> => {
    const response = await apiClient.post<NotificationChannel>('/api/notifications/channels', data);
    return response.data;
  },

  updateChannel: async (id: string, data: NotificationChannelUpdate): Promise<NotificationChannel> => {
    const response = await apiClient.patch<NotificationChannel>(`/api/notifications/channels/${id}`, data);
    return response.data;
  },

  deleteChannel: async (id: string): Promise<void> => {
    await apiClient.delete<void>(`/api/notifications/channels/${id}`);
  },

  testChannel: async (id: string, payload: ChannelTestRequest): Promise<ChannelTestResponse> => {
    const response = await apiClient.post<ChannelTestResponse>(`/api/notifications/channels/${id}/test`, payload);
    return response.data;
  },

  // Templates
  listTemplates: async (): Promise<NotificationTemplate[]> => {
    const response = await apiClient.get<NotificationTemplate[]>('/api/notifications/templates');
    return response.data;
  },

  createTemplate: async (data: NotificationTemplateCreate): Promise<NotificationTemplate> => {
    const response = await apiClient.post<NotificationTemplate>('/api/notifications/templates', data);
    return response.data;
  },

  updateTemplate: async (id: string, data: NotificationTemplateUpdate): Promise<NotificationTemplate> => {
    const response = await apiClient.patch<NotificationTemplate>(`/api/notifications/templates/${id}`, data);
    return response.data;
  },

  deleteTemplate: async (id: string): Promise<void> => {
    await apiClient.delete<void>(`/api/notifications/templates/${id}`);
  },

  // Recipients
  listRecipients: async (): Promise<NotificationRecipient[]> => {
    const response = await apiClient.get<NotificationRecipient[]>('/api/notifications/recipients');
    return response.data;
  },

  createRecipient: async (data: NotificationRecipientCreate): Promise<NotificationRecipient> => {
    const response = await apiClient.post<NotificationRecipient>('/api/notifications/recipients', data);
    return response.data;
  },

  updateRecipient: async (id: string, data: NotificationRecipientUpdate): Promise<NotificationRecipient> => {
    const response = await apiClient.patch<NotificationRecipient>(`/api/notifications/recipients/${id}`, data);
    return response.data;
  },

  deleteRecipient: async (id: string): Promise<void> => {
    await apiClient.delete<void>(`/api/notifications/recipients/${id}`);
  },

  // Rules
  listRules: async (): Promise<NotificationRule[]> => {
    const response = await apiClient.get<NotificationRule[]>('/api/notifications/rules');
    return response.data;
  },

  createRule: async (data: NotificationRuleCreate): Promise<NotificationRule> => {
    const response = await apiClient.post<NotificationRule>('/api/notifications/rules', data);
    return response.data;
  },

  updateRule: async (id: string, data: NotificationRuleUpdate): Promise<NotificationRule> => {
    const response = await apiClient.patch<NotificationRule>(`/api/notifications/rules/${id}`, data);
    return response.data;
  },

  deleteRule: async (id: string): Promise<void> => {
    await apiClient.delete<void>(`/api/notifications/rules/${id}`);
  },

  testRule: async (id: string, payload: RuleTestRequest): Promise<RuleTestResponse> => {
    const response = await apiClient.post<RuleTestResponse>(`/api/notifications/rules/${id}/test`, payload);
    return response.data;
  },

  // Delivery logs
  listLogs: async (filters?: DeliveryLogFilters): Promise<DeliveryLog[]> => {
    const response = await apiClient.get<DeliveryLog[]>('/api/notifications/logs', {
      params: filters,
    });
    return response.data;
  },
};
